rustls/src/server/handy.rs

```diff
@@ -46,32 +45,51 @@ mod cache {
         /// Make a new ServerSessionMemoryCache.  `size` is the maximum
         /// number of stored sessions, and may be rounded-up for
         /// efficiency.
+        #[cfg(feature = "std")]
         pub fn new(size: usize) -> Arc<Self> {
             Arc::new(Self {
                 cache: Mutex::new(limited_cache::LimitedCache::new(size)),
             })
         }
+
+        /// Make a new ServerSessionMemoryCache.  `size` is the maximum
+        /// number of stored sessions, and may be rounded-up for
+        /// efficiency.
+        #[cfg(not(feature = "std"))]
+        pub fn new<M: crate::lock::MakeMutex>(size: usize) -> Arc<Self> {
+            Arc::new(Self {
+                cache: M::make_mutex(limited_cache::LimitedCache::new(size)),
+            })
+        }
     }
 
     impl server::StoresServerSessions for ServerSessionMemoryCache {
         fn put(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
-            self.cache
-                .lock()
-                .unwrap()
-                .insert(key, value);
+            #[cfg(feature = "std")]
+            let mut cache = self.cache.lock().unwrap();
+            #[cfg(not(feature = "std"))]
+            let mut cache = self.cache.lock();
+
+            cache.insert(key, value);
             true
         }
 
         fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
-            self.cache
-                .lock()
-                .unwrap()
-                .get(key)
-                .cloned()
+            #[cfg(feature = "std")]
+            let cache = self.cache.lock().unwrap();
```

Comment:
If possible, it would be nicer if the `Lock` trait impls can abstract over poisoning.

Reply:
> it would be nicer if the Lock trait impls can abstract over poisoning.

@djc I agree it would be nicer, but so far the ideas I've had to do it don't seem like a net positive. Did you have a specific approach in mind? 

My thoughts so far:

The `Lock` trait isn't in-play for the std-enabled case and I don't think we want to change that to avoid overhead in the lock/unlock path on std builds. When std is enabled the locks are `std::sync::Mutex` re-exports, not an arc over a `dyn Lock`. I think that rules out making the std case look like the no-std case for calling `lock()`.

That leaves making the no-std `lock()` look more like the std case by introducing the handling of a poisoned result. We could make the `Lock` trait's `lock()` fn return `Result<MutexGuard<'_, Self::Data>, Self::PoisonError>;` instead of `MutexGuard<'_, Self::Data>` but adding a second generic parameter to the `Lock` trait for the error will make the no-std usages even more divergent from the std-usages. Fixing the type to a concrete `Error` and not making it generic might be one option.

Am I missing/overthinking something here?

Reply:
> How about we just yield Option? Seems to be what we already use in the ticketer in practice?

Went with a wrapper type that has a lock fn yielding an `Option`.